databuilder/query_model/transforms.py

```diff
@@ -168,6 +174,21 @@ def make_sortable(col):
     return col
 
 
+def collapse_statically_known_non_nulls(node, reverse_index):
+    """
+    Some operations never return null. When such operations are nested inside IsNull, the result
+    is guaranteed to be False; so we can collapse an entire subtree to a single value.
+
+    This allows us to avoid some expressions which would result in SQL that exceeds databases'
+    limits on nested CASE statements.
+    """
+    if isinstance(node.source, Function.IsNull):
+        del reverse_index[node.source]
```

Comment:
Not sure if this is really necessary. It seems like tidying up is a good idea, but we're not doing anything to prune the now-garbage nodes from the list of nodes under consideration, so it's rather cosmetic.

Reply:
(It's also not tested.)

Reply:
It seems reasonable to me too, but I think I'm also prone to bits of purely cosmetic tidying so I'm maybe not the best sounding board!